StreamChatClient_v3/ChatClient.swift

```diff
@@ -151,80 +142,264 @@ public class Client<ExtraData: ExtraDataTypes> {
+    /// is called with an error.
+    public func setAnonymousUser(shouldConnectAutomatically: Bool = true, completion: ((Error?) -> Void)? = nil) {
+        disconnect()
+        prepareEnvironmentForNewUser(userId: .anonymous, role: .anonymous, extraData: nil) { error in
+            guard error == nil else {
+                completion?(error)
+                return
+            }
+            
+            self.connect(completion: completion)
+        }
+    }
+    
+    public func setGuestUser(userId: UserId, extraData: ExtraData.User? = nil, completion: ((Error?) -> Void)? = nil) {
+        fatalError("not implemented")
+    }
+    
+    /// Sets a new current user of the ChatClient.
+    ///
+    /// - Important: Setting a new user disconnects all the existing controllers. You should create new controllers
+    /// if you want to keep receiving updates about the newly set user.
+    ///
+    /// - Parameters:
+    ///   - userId: The id of the new current user.
+    ///   - userExtraData: You can optionally provide additional data to be set for the user. This is an equivalent of
+    ///   setting the current user detail data manually using `CurrentUserController`.
+    ///   - token: You can provide a token which is used for user authentication. If the `token` is not explicitly provided,
+    ///   the client uses `ChatClientConfig.tokenProvider` to obtain a token. If you haven't specified the token provider,
+    ///   providing a token explicitly is required. In case both the `token` and `ChatClientConfig.tokenProvider` is specified,
+    ///   the `token` value is used.
+    ///   - completion: Called when the new user is successfully set.
+    public func setUser(userId: UserId,
+                        userExtraData: ExtraData.User? = nil,
+                        token: Token? = nil,
+                        completion: ((Error?) -> Void)? = nil) {
+        guard token != nil || config.tokenProvider != nil else {
+            log.assert(false,
+                       "The provided token is `nil` and `ChatClientConfig.tokenProvider` is also `nil`. You must either provide " +
+                           "a token explicitly or set `TokenProvider` in `ChatClientConfig`.")
+            completion?(ClientError.MissingToken())
+            return
+        }
+        
+        guard userId != currentUserId else {
+            log.warning("New user with id:<\(userId)> is not set because it's similar to the already logged-in user.")
+            completion?(nil)
+            return
+        }
+        
+        disconnect()
+        
+        prepareEnvironmentForNewUser(userId: userId, role: .user, extraData: userExtraData) { error in
+            guard error == nil else {
+                completion?(error)
+                return
+            }
+            
+            if let token = token {
+                self.currentToken = token
```

Comment:
Why don't we:
```swift
if let token = token {
  config.tokenProvider = { $2(token) }
}
```
and treat the rest of the code as if only tokenProvider is provided?

Reply:
Hmm, the current setup allows you to override the provider with a token if you need it. I like your suggestion for its cleverness but I'd stay to the more boring implementation for this :)